In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [5]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
market_data_scaled = StandardScaler().fit_transform(df_market_data[df_market_data.columns])

In [6]:
# Create a DataFrame with the scaled data
market_data_scaled_df = pd.DataFrame(market_data_scaled, columns=df_market_data.columns)

# Copy the crypto names from the original DataFrame
market_data_scaled_df["cryptos"] = df_market_data.index

# Set the coin_id column as index
market_data_scaled_df.set_index("cryptos", inplace=True)

# Display the scaled DataFrame
market_data_scaled_df

price_change_percentage_24h  price_change_percentage_7d  \
cryptos                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
cryptos                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [7]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

# Create an empty list to store the inertia values
inertia = []


In [8]:
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k:
	elbow_model = KMeans(n_clusters=i, random_state=1)
	elbow_model.fit(market_data_scaled_df)
	inertia.append(elbow_model.inertia_)

c:\Users\Jitesh Makan\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\Jitesh Makan\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\Jitesh Makan\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\Jitesh Makan\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to 

In [9]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {
	"k": k,
	"inertia": inertia
}

# Create a DataFrame with the data to plot the Elbow curve
elbow_data_df = pd.DataFrame(elbow_data)
elbow_data_df

k     inertia
0   1  287.000000
1   2  212.123342
2   3  165.136752
3   4   79.022435
4   5   66.413051
5   6   61.634088
6   7   46.074467
7   8   43.392694
8   9   32.913223
9  10   30.469111

In [10]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_data_df.hvplot.line(
	x = "k",
	y = "inertia",
	title = "Elbow Chart for Crypto Market Data",
	xticks = k
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** An examination of the Elbow Chart derived from the cryptocurrency market dataset indicates that k = 4 is the optimal number of clusters. This conclusion is supported by the characteristic inflection point observed in the curve, beyond which additional clusters yield only marginal reductions in within-cluster sum of squares (WCSS). The diminishing rate of improvement beyond k = 4 suggests that further increases in the number of clusters offer minimal representational or structural benefit, reinforcing k = 4 as a methodologically sound choice for capturing the inherent groupings within the data.

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [11]:
# Initialize the K-Means model using the best value for k
k_model = KMeans(n_clusters=4, random_state=1)

In [12]:
# Fit the K-Means model using the scaled DataFrame
k_model.fit(market_data_scaled_df)

c:\Users\Jitesh Makan\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=1)

In [13]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
k_prediction = k_model.predict(market_data_scaled_df)

# Print the resulting array of cluster values.
k_prediction

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 3, 2, 0, 0, 1, 0, 0, 0, 0])

In [14]:
# Create a copy of the scaled DataFrame
market_data_clustered = market_data_scaled_df.copy()

In [15]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
market_data_clustered['k_clusters'] = k_prediction

# Display the copy of the scaled DataFrame
market_data_clustered

price_change_percentage_24h  price_change_percentage_7d  \
cryptos                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
cryptos                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

In [16]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
market_data_clustered.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by='k_clusters',
    hover_cols=['cryptos'],
    title = "K-Means Clustering of Cryptocurrencies"
)

:NdOverlay   [k_clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,cryptos)

---

### Optimize Clusters with Principal Component Analysis.

In [17]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [18]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
market_data_pca = pca.fit_transform(market_data_scaled_df)

# View the scaled PCA data
market_data_pca[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [19]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
display(pca.explained_variance_ratio_)
display(pca.explained_variance_ratio_.sum())


array([0.3719856 , 0.34700813, 0.17603793])

0.8950316570309841

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance captured by the three retained principal components is 0.895, indicating that approximately 89.5% of the original data's variability is preserved post-transformation. While this level of variance retention is generally acceptable for many applications, it may be methodologically insufficient for modeling highly volatile datasets—such as those derived from cryptocurrency financial markets—where lower-variance components can still contain critical information about short-term fluctuations or anomalous behavior. The exclusion of the remaining ~10.5% of variance introduces a risk of information loss that could impair the performance of subsequent analyses, particularly those sensitive to minor but meaningful variations in the data. This highlights the importance of carefully evaluating the trade-off between dimensionality reduction and information retention, especially in contexts characterized by high complexity and unpredictability.

In [20]:
# Create a new DataFrame with the PCA data.
market_data_pca_df = pd.DataFrame(
    market_data_pca,
    columns=['PCA1','PCA2', 'PCA3'])

# Copy the crypto names from the original scaled DataFrame
market_data_pca_df['cryptos'] = market_data_scaled_df.index

# Set the coin_id column as index
market_data_pca_df.set_index('cryptos', inplace=True)

# Display the scaled PCA DataFrame
market_data_pca_df

PCA1      PCA2      PCA3
cryptos                                           
bitcoin              -0.600667  0.842760  0.461595
ethereum             -0.458261  0.458466  0.952877
tether               -0.433070 -0.168126 -0.641752
ripple               -0.471835 -0.222660 -0.479053
bitcoin-cash         -1.157800  2.041209  1.859715
binancecoin          -0.516534  1.388377  0.804071
chainlink            -0.450711  0.517699  2.846143
cardano              -0.345600  0.729439  1.478013
litecoin             -0.649468  0.432165  0.600303
bitcoin-cash-sv      -0.759014 -0.201200 -0.217653
crypto-com-chain     -0.248198 -1.376252 -1.462026
usd-coin             -0.438408 -0.175337 -0.663388
eos                  -0.693425 -0.473815 -0.527597
monero                0.060499  2.909404  1.498571
tron                 -0.393352 -0.108192 -0.012756
tezos                -0.796176 -0.494409  1.082812
okb                   0.064075 -1.269825 -1.098829
stellar              -0.489015 -0.732719 -0.062543
cosmos               -0.306272  0.703415  1.714224
cdai                 -0.513528 -0.142802 -0.656566
neo                  -0.362120 -0.986914 -0.728752
wrapped-bitcoin      -0.604265  0.827398  0.439316
leo-token            -0.413296 -0.674115 -1.076628
huobi-token          -0.407483 -0.212507 -0.351426
nem                   0.608974  0.563532 -1.148742
binance-usd          -0.450211 -0.151019 -0.647401
iota                 -0.764665 -0.517886  0.204990
vechain              -0.556315 -1.938209 -1.261776
zcash                -0.425147  0.492976  1.058048
theta-token           2.676868 -0.013954 -1.965207
dash                 -0.613923 -0.479337  0.339565
ethereum-classic     -0.579924 -0.356334 -0.114942
ethlend               8.089018 -3.896891  2.301382
maker                -0.389045  0.165041  0.379414
havven                0.865762 -2.261882  0.275583
omisego               0.111675  0.428316 -1.205398
celsius-degree-token  4.792395  6.767679 -1.986985
ontology             -0.632355 -2.108117 -0.652227
ftx-token            -0.593142  0.021485  0.209911
true-usd             -0.458131 -0.135734 -0.635284
digibyte             -0.297910 -0.191126 -0.909602

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [21]:
# Create a list with the number of k-values from 1 to 11
pca_k = list(range(1,11))

# Create an empty list to store the inertia values
pca_inertia = []

In [22]:
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in pca_k:
    model_pca = KMeans(n_clusters=i, random_state=1)
    model_pca.fit(market_data_pca_df)
    pca_inertia.append(model_pca.inertia_)

c:\Users\Jitesh Makan\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\Jitesh Makan\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\Jitesh Makan\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\Jitesh Makan\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to 

In [23]:
# Create a dictionary with the data to plot the Elbow curve
pca_elbow_data = {
    "k":pca_k,
    "inertia":pca_inertia
}

# Create a DataFrame with the data to plot the Elbow curve
pca_elbow_df = pd.DataFrame(pca_elbow_data)

pca_elbow_df

k     inertia
0   1  256.874086
1   2  182.339530
2   3  135.442408
3   4   49.665497
4   5   38.672582
5   6   34.507577
6   7   23.728365
7   8   18.574523
8   9   15.572672
9  10   11.407667

In [24]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
pca_elbow_df.hvplot.line(
    x= 'k',
    y = 'inertia',
    xticks = pca_k,
    title = "Elbow Curve"
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:**An analysis of the Elbow curve derived from the PCA-transformed dataset indicates that an optimal choice for the number of clusters, k, is 4. Beyond this point, the marginal gain in explained variance or reduction in within-cluster inertia becomes negligible. This plateau suggests diminishing returns in clustering performance for higher values of k, reinforcing the selection of k = 4 as a balanced point that captures the essential structure of the data without overfitting.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No significant difference was observed between the Elbow Curve charts generated with and without the application of PCA. Visually, the curves appear nearly identical, indicating that PCA preserves the essential clustering structure of the original dataset. This consistency underscores the efficiency and effectiveness of PCA as a preprocessing step for identifying optimal k values, particularly in high-dimensional or large-scale datasets where computational complexity and noise may otherwise obscure underlying patterns.

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [25]:
# Initialize the K-Means model using the best value for k
pca_kmodel = KMeans(n_clusters=4, random_state=1)

In [26]:
# Fit the K-Means model using the PCA data
pca_kmodel.fit(market_data_pca_df)

c:\Users\Jitesh Makan\anaconda3\Lib\site-packages\sklearn\cluster\_kmeans.py:1429: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=1)

In [27]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
pca_k_predictions = pca_kmodel.predict(market_data_pca_df)

# Print the resulting array of cluster values.
pca_k_predictions

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 3, 2, 0, 0, 1, 0, 0, 0, 0])

In [28]:
# Create a copy of the scaled PCA DataFrame
market_data_pca_clustered = market_data_pca_df.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
market_data_pca_clustered['pca_clusters'] = pca_k_predictions

# Display the copy of the scaled PCA DataFrame
market_data_pca_clustered.head()

PCA1      PCA2      PCA3  pca_clusters
cryptos                                                 
bitcoin      -0.600667  0.842760  0.461595             2
ethereum     -0.458261  0.458466  0.952877             2
tether       -0.433070 -0.168126 -0.641752             0
ripple       -0.471835 -0.222660 -0.479053             0
bitcoin-cash -1.157800  2.041209  1.859715             2

In [29]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
market_data_pca_clustered.hvplot.scatter(
    x = "PCA1",
    y = "PCA2",
    by = "pca_clusters",
    hover_cols=["cryptos"],
    title = "PCA Clustering Results"
)


:NdOverlay   [pca_clusters]
   :Scatter   [PCA1]   (PCA2,cryptos)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [30]:
# Composite plot to contrast the Elbow curves
plot1 = elbow_data_df.hvplot.line(x='k', y='inertia', xticks=k, label="K-Means Clustering")
plot2 = pca_elbow_df.hvplot.line(x='k', y='inertia', xticks=k, label="PCA Clustering")

composite_plot1 = plot1 + plot2

composite_plot1

:Layout
   .Curve.K_hyphen_minus_Means_Clustering :Curve   [k]   (inertia)
   .Curve.PCA_Clustering                  :Curve   [k]   (inertia)

In [31]:
# Composite plot to contrast the Elbow curves
plot1 = elbow_data_df.hvplot.line(x='k', y='inertia', xticks=k, label="regular plot")
plot2 = pca_elbow_df.hvplot.line(x='k', y='inertia', xticks=k, label="pca plot")

composite_plot1 = plot1 * plot2.opts(title='Regular vs PCA Data Elbow Method Comparison')

composite_plot1

:Overlay
   .Curve.Regular_plot :Curve   [k]   (inertia)
   .Curve.Pca_plot     :Curve   [k]   (inertia)

In [32]:
plot3 = market_data_clustered.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by='k_clusters',
    hover_cols=["Crypto Name", "cryptos"],
    label="K-Means Clustering",
    s=80,  # marker size
    cmap='Category10'
)

plot4 = market_data_pca_clustered.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    by="pca_clusters",
    hover_cols=["Crypto Name", "cryptos"],
    label="PCA Clustering",
    s=80,  # marker size
    cmap='Set1'  # contrasting color map
)

composite_plot3 = plot3 + plot4

composite_plot3


:Layout
   .NdOverlay.K_hyphen_minus_Means_Clustering :NdOverlay   [k_clusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,cryptos)
   .NdOverlay.PCA_Clustering                  :NdOverlay   [pca_clusters]
      :Scatter   [PCA1]   (PCA2,cryptos)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:**Principal Component Analysis (PCA), while primarily employed for dimensionality reduction, also enhances the interpretability of clustering by amplifying variance-based distinctions within the dataset. This effect is particularly evident when examining the clustering results in the PCA-reduced feature space. Notably, clusters 1 and 3 each consist of a single data point that is clearly separated from the more compact and cohesive clusters 0 and 2. This separation is more visually pronounced post-PCA, whereas in the original high-dimensional space, cluster 1 appears embedded within the spatial domain of cluster 0, despite its distinct cluster assignment. The alignment between the PCA’s explained variance ratios and the Elbow Method further validates the dimensionality selection: a reduction in the number of features results in a marked increase in total inertia, highlighting the loss of structural fidelity beyond a certain threshold. Collectively, these observations underscore PCA's utility not only in preserving variance but also in revealing latent structure and outlier behavior that may be less discernible in the original feature space.